Cython/Compiler/Optimize.py

```diff
@@ -228,6 +228,12 @@ def _optimise_for_loop(self, node, iterable, reversed=False):
             return self._transform_bytes_iteration(node, iterable, reversed=reversed)
         if iterable.type is Builtin.unicode_type:
             return self._transform_unicode_iteration(node, iterable, reversed=reversed)
+        if iterable.type in (Builtin.list_type, Builtin.tuple_type):
+            return self._transform_indexable_iteration(node, iterable, reversed=reversed)
```

Comment:
There is already special-casing for these, both in the general "don't know the type" case as well as in the case that we know the type of the iterable to be tuple or list. Not sure which is better, but my guess is that at least the current code that this transformation produces isn't as tight as what we already generate.

That doesn't mean that it may not be worth removing the current special-casing, just that we shouldn't introduce a regression along the way.

Reply:
The code from this PR currently likes to juggle every result between about 3 different unnecessary temps. I suspect that can be improved though.

Mostly, this was in there because it seemed like an easy place to start and check the general working; it can obviously be removed if it makes things worse.